project-koku/koku
# COST-2886: (2887-2888): Add exclusions to our endpoints.

## Jira Ticket

[COST-2886](https://issues.redhat.com/browse/COST-2886)
[COST-2887](https://issues.redhat.com/browse/COST-2887)
[COST-2888](https://issues.redhat.com/browse/COST-2888)

## Description

This pull request will add exclusions to QueryParams object's parameters attribute.

## Testing

1. Bring up the project with gunicorn

2. Go to this page: 
- http://localhost:4242/#/features/strategies/cost-enable-negative-filtering

3. Add the "schema" `acct1234567`

4. Add a source. 

5. This pull request will need to go in first: 
- https://github.com/project-koku/koku/pull/3869

6. Then you can hit an endpoint like this:
```
http://localhost:8000/api/cost-management/v1/reports/gcp/costs/?filter[time_scope_units]=month&filter[resolution]=monthly&group_by[gcp_project]=*&filter[time_scope_value]=-1&exclude[gcp_project]=cost-management-dev
```

7. Then check the logs:
```
koku_server       | [2022-09-14 17:32:30,153] INFO None enable_negative_filtering context: {'schema': 'acct1234567'}
koku_server       | [2022-09-14 17:32:30,154] INFO None     Negative Filtering Enabled acct1234567
koku_server       | [2022-09-14 17:32:30,179] INFO None Setting self.query_exclusions: (OR: ('project_name__icontains', 'cost-management-dev'), ('project_id__icontains', 'cost-management-dev'))
```


## TODO


## Examples:

[Azure grouped by service](http://localhost:8000/api/cost-management/v1/reports/azure/costs/?filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly&group_by[service_name]=*)
<img width="1251" alt="Screen Shot 2022-09-26 at 1 38 09 PM" src="https://user-images.githubusercontent.com/16995455/192343990-9ae599ae-5a45-4d9f-bdd0-3766b13a1452.png">
[azure_exclude.txt](https://github.com/project-koku/koku/files/9648463/azure_exclude.txt)


[Azure grouped by service with excludes](http://localhost:8000/api/cost-management/v1/reports/azure/costs/?filter[time_scope_value]=-1&filter[time_scope_units]=month&filter[resolution]=monthly&group_by[service_name]=*&exclude[service_name]=Storage)
<img width="1206" alt="Screen Shot 2022-09-26 at 1 38 56 PM" src="https://user-images.githubusercontent.com/16995455/192344015-260c6429-5f28-4793-8918-e2755b9cf8cf.png">
[azure.txt](https://github.com/project-koku/koku/files/9648476/azure.txt)


Additional Examples:
- [gcp_group_with_multi_exclude.txt](https://github.com/project-koku/koku/files/9649063/gcp_group_with_multi_exclude.txt) (Note I excluded both projects we had in the group by so we do expect zero here.)
- [gcp_group_by_one_exclude.txt](https://github.com/project-koku/koku/files/9649050/gcp_group_by_one_exclude.txt)
- [gcp_group_by.txt](https://github.com/project-koku/koku/files/9649006/gcp_group_by.txt)
- [multi_tag_exclude_example.txt](https://github.com/project-koku/koku/files/9648396/multi_tag_exclude_example.txt)



...
